File: backend/chat/app/core/providers/google_provider.py
import os
import google.generativeai as genai
from app.core.providers.base_provider import BaseAIProvider


class GoogleProvider(BaseAIProvider):
    def __init__(self):
        # Configura a API Key do Gemini
        genai.configure(api_key="AIzaSyAeGuXWC0THG2rAAGcJePv3W_cqLW90iqA")
        # Modelo principal de chat
        self.chat_model = genai.GenerativeModel("gemini-2.5-flash-lite")

    def embed_text(self, text: str) -> list[float]:
        """Gera embedding usando o modelo de embeddings do Gemini."""
        result = genai.embed_content(
            model="models/embedding-001",
            content=text,
        )
        return result["embedding"]

    def ask_question(self, question: str, context: str) -> str:
        """Gera resposta usando o modelo Gemini."""
        prompt = f"""
        Você é um assistente de nutrição amigável e informativo.
        Responda usando apenas o contexto abaixo.
        Se não souber, diga "Não tenho informações sobre isso no meu banco de dados nutricional".

        CONTEXTO:
        {context}

        PERGUNTA:
        {question}
        """
        return self.chat(prompt)

    def chat(self, prompt: str) -> str:
        """Envia um prompt genérico ao Gemini."""
        response = self.chat_model.generate_content(prompt)
        return response.text

File: backend/chat/src/core/providers/GoogleProvider.ts
import {
  BatchEmbedContentsRequest,
  BatchEmbedContentsResponse,
  ContentEmbedding, EmbedContentRequest,
  GenerativeModel,
  GoogleGenerativeAI
} from '@google/generative-ai';
import BaseAIProvider from './BaseAIProvider.js';
import path from "path";
import fs from "fs";

class GoogleProvider extends BaseAIProvider {
  client: any;
  chatModel: any;
  embeddingModel: GenerativeModel;

  constructor() {
    super();
    const apiKey = process.env.GOOGLE_API_KEY;
    if (!apiKey) {
      throw new Error('GOOGLE_API_KEY não está definida');
    }
    this.client = new GoogleGenerativeAI(apiKey);
    this.chatModel = this.client.getGenerativeModel({ model: 'gemini-2.5-flash-lite' });
    this.embeddingModel = this.client.getGenerativeModel({ model: 'gemini-embedding-001' });
  }

  async embedChunks(texts: string[]) {
    try {
      // O SDK oficial aceita um objeto com a propriedade requests
      const result = await this.embeddingModel.batchEmbedContents({
        requests: texts.map((t) => ({
          content: { role: 'user', parts: [{ text: t }] },
        })),
      });

      // Retorna apenas os arrays de valores numéricos
      return result.embeddings.map(e => e.values);
    } catch (error: any) {
      console.error("Erro no Batch Embedding:", error);
      throw new Error(`Erro ao gerar batch embedding: ${error.message}`);
    }
  }

  async askQuestion(history: string, context: string, question: string) {

    const promptPath = path.join(__dirname, '..', 'prompts', 'system_prompt.md');
    let fullPrompt = fs.readFileSync(promptPath, 'utf8');

    fullPrompt = fullPrompt
        .replace('{{historyStr}}', history)
        .replace('{{context}}', context)
        .replace('{{question}}', question);

    return this.chat(fullPrompt);
  }

  async chat(prompt: any) {
    try {
      const response = await (this as any).chatModel.generateContent(prompt);
      return response.response.text();
    } catch (error: any) {
      throw new Error(`Erro ao gerar resposta do chat: ${error.message}`);
    }
  }
}

export default GoogleProvider;

